google-ads/src/main/java/com/google/ads/googleads/lib/GoogleAdsClient.java

```diff
@@ -149,16 +168,93 @@ public static Builder newBuilder() {
+      boolean hasServiceAccountKeys =
+          SERVICE_ACCOUNT_OAUTH_KEYS.stream().anyMatch(k -> k.getPropertyValue(properties) != null);
+      if (!(hasInstalledAppKeys || hasServiceAccountKeys)) {
+        // Entries missing for both types of credentials.
+        throw new IllegalArgumentException(
+            String.format(
+                "Missing properties for credentials. Please modify properties to include entries"
+                    + " for %s if using installed application/web flow credentials, or %s if using"
+                    + " service account credentials.",
+                INSTALLED_APP_OAUTH_KEYS, SERVICE_ACCOUNT_OAUTH_KEYS));
+      } else if (hasInstalledAppKeys && hasServiceAccountKeys) {
+        // Entries specified for both types of credentials.
+        throw new IllegalArgumentException(
+            String.format(
+                "Entries found in properties for both %s and %s. Please modify properties to either"
+                    + " include entries for %s if using installed application/web flow credentials,"
+                    + " or %s if using service account credentials.",
+                ConfigPropertyKey.SERVICE_ACCOUNT_SECRETS_PATH,
+                ConfigPropertyKey.REFRESH_TOKEN,
+                INSTALLED_APP_OAUTH_KEYS,
+                SERVICE_ACCOUNT_OAUTH_KEYS));
+      }
+
+      // Constructs the credentials object using the appropriate set of properties.
+      setCredentials(
+          hasInstalledAppKeys
+              ? createUserCredentials(properties)
+              : createServiceAccountCredentials(properties));
+    }
+
+    /**
+     * Creates and returns a {@link Credentials} object based on the installed app or web server
+     * configuration keys.
+     */
+    private Credentials createUserCredentials(Properties properties) {
+      String refreshToken = ConfigPropertyKey.REFRESH_TOKEN.getPropertyValue(properties);
+      Preconditions.checkNotNull(
+          "No refresh token specified under the key: %s", ConfigPropertyKey.REFRESH_TOKEN);
+      String clientId = ConfigPropertyKey.CLIENT_ID.getPropertyValue(properties);
+      Preconditions.checkNotNull(
+          "No client ID specified under the key: %s", ConfigPropertyKey.CLIENT_ID);
+      String clientSecret = ConfigPropertyKey.CLIENT_SECRET.getPropertyValue(properties);
+      Preconditions.checkNotNull(
+          "No client secret specified under the key: %s", ConfigPropertyKey.CLIENT_SECRET);
+
+      return UserCredentials.newBuilder()
+          .setClientId(ConfigPropertyKey.CLIENT_ID.getPropertyValue(properties))
+          .setClientSecret(ConfigPropertyKey.CLIENT_SECRET.getPropertyValue(properties))
+          .setRefreshToken(refreshToken)
+          .build();
+    }
+
+    /**
+     * Creates and returns a {@link Credentials} object based on the service account configuration
+     * keys.
+     */
+    private Credentials createServiceAccountCredentials(Properties properties) throws IOException {
```

Comment:
It seems like this could be of general use to external users. What do you think about adding a `public static` method which takes `Strings` and returns `Credentials`?

Reply:
I'd prefer to leave that to the OAuth library. The `ServiceAccountCredentials` class already has factory and builder methods.

Were you thinking the value-add would be specifying the scope? If so, I could make the `GOOGLE_ADS_API_SCOPE` constant `public`.

Reply:
Do you know if the recent changes to mark the scope as "documented" mean it will soon appear in generated code? I just want to be sure I don't add a public constant only to remove it shortly thereafter.

Reply:
Adam should know for sure.

Reply:
Just confirmed that with that change, we'll be able to get this constant from generated code, so I'll leave this private for now and we can drop this constant altogether in an upcoming release.

@nwbirnie , sound good?

Reply:
I guess all the libraries will have to clean this up, though the exact form of how this const looks like is something I haven't looked at. Perhaps we can pickup this cleanup as an item for Q1?

Reply:
SGTM - let's add this for Q1.

Reply:
+1. I'll mark this conversation resolved and ping Knack about adding this to an upcoming release in Q1.